storage-proofs-update/src/circuit.rs

```diff
@@ -0,0 +1,587 @@
+                AllocatedNum::alloc(
+                    cs.namespace(|| format!("partition_path_sibling_{}", i)),
+                    || siblings[0].ok_or(SynthesisError::AssignmentMissing),
+                )
+                .map(|sibling| vec![sibling])
+            })
+            .collect::<Result<Vec<Vec<AllocatedNum<Fr>>>, SynthesisError>>()?;
+
+        // Assert that the witnessed `comm_r_last_old` and `comm_r_last_new` are consistent with the
+        // public `comm_r_old` and `comm_r_new` via `comm_r = H(comm_c || comm_r_last)`.
+        let comm_r_old_calc = <TreeR::Hasher as Hasher>::Function::hash2_circuit(
+            cs.namespace(|| "comm_r_old_calc"),
+            &comm_c,
+            &comm_r_last_old,
+        )?;
+        cs.enforce(
+            || "enforce comm_r_old_calc == comm_r_old",
+            |lc| lc + comm_r_old_calc.get_variable(),
+            |lc| lc + CS::one(),
+            |lc| lc + comm_r_old.get_variable(),
+        );
+        let comm_r_new_calc = <TreeR::Hasher as Hasher>::Function::hash2_circuit(
+            cs.namespace(|| "comm_r_new_calc"),
+            &comm_c,
+            &comm_r_last_new,
+        )?;
+        cs.enforce(
+            || "enforce comm_r_new_calc == comm_r_new",
+            |lc| lc + comm_r_new_calc.get_variable(),
+            |lc| lc + CS::one(),
+            |lc| lc + comm_r_new.get_variable(),
+        );
+
+        // Assert that this partition's `apex_leafs` are  consistent with the public `comm_d_new`.
+        apex_por(
+            cs.namespace(|| "apex_gadget"),
+            apex_leafs.clone(),
+            partition_bits.clone(),
+            partition_path,
+            comm_d_new,
+        )?;
+
+        // Generate `challenge_sans_partition_bit_len` number of random bits for each challenge.
+        // For each challenge generate a random index in `0..number of leaf's per partition`; we
+        // append the partition-index's bits onto the random bits generated for each challenge
+        // producing a challenge in `0..sector_nodes` which is guaranteed to lie within this
+        // partition's subset of leafs.
+        let challenge_sans_partition_bit_len = challenge_bit_len - partition_bit_len;
+        let generated_bits = gen_challenge_bits(
+            cs.namespace(|| "gen_challenge_bits"),
+            &comm_r_new,
+            &partition,
+            challenge_count,
+            challenge_sans_partition_bit_len,
+        )?;
+
+        for (c_index, c_bits_without_partition) in generated_bits.into_iter().enumerate() {
+            let c_bits: Vec<AllocatedBit> = c_bits_without_partition
+                .iter()
+                .chain(partition_bits.iter())
+                .cloned()
+                .collect();
+
+            // Compute this challenge's `rho`.
+            let c_high = get_challenge_high_bits(
+                cs.namespace(|| format!("get_challenge_high_bits (c_index={})", c_index)),
+                &c_bits_without_partition,
```

Comment:
c_high for `rho` generation should include partition bits, I think we misscommunicated.